### ISSUE TITLE
Support ml_train_datagen_cb() for C#

### ISSUE THREAD

**Issue by songgot** [nnstreamer/nntrainer#1948]
# We need some interface to support multi input_num  and solve double pointer issue.

It seems that **input and **label of our datagen callback cannot be supported by C# Interop.

C# only receive a **pointer as IntPtr(single pointer). so I tried the following.
As a first attempt, I tried to process it as a one-dimensional array.
< Interop.Dataset.cs >
~~~

        /* typedef int (*ml_train_datagen_cb)(float **input, float **label, bool *last, void *user_data) */
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        internal delegate void DatagenCallback(IntPtr input, IntPtr label, IntPtr last, IntPtr userData);

      /* int ml_train_dataset_add_generator(ml_train_dataset_h dataset, ml_train_dataset_mode_e mode, ml_train_datagen_cb cb, void *user_data) */
        [DllImport(Libraries.Nntrainer, EntryPoint ="ml_train_dataset_add_generator", CallingConvention = CallingConvention.Cdecl)]
        internal static extern NNTrainerError AddGenerator(IntPtr datasetHandle, NNTrainerDatasetMode mode, DatagenCallback callback, IntPtr userData);
~~~
<DatagenEventArgs.cs>
~~~
public class DatagenEventArgs : EventArgs
{
        public float[] Input { get; set;}
        public float[] Label { get; set;}
        public float[] Last {get; set;}
}
~~~
<Dataset.cs>
~~~
        public void AddGenerator(NNTrainerDatasetMode mode, IntPtr userData)
        {
            datagenCallback = (intput, label, last, _) =>
            {
                var args = new DatagenEventArgs();
                eventReceived?.Invoke(this, args);

                Marshal.Copy(args.Input, 0, intput, args.Input.Length);
                Marshal.Copy(args.Label, 0, label, args.Label.Length);
                Marshal.Copy(args.Last, 0, last, args.Label.Length);
            };

            NNTrainerError ret = Interop.Dataset.AddGenerator(handle, mode, datagenCallback, userData);
            NNTrainer.CheckException(ret, "Failed to set DatagenCallback");
            Log.Info(NNTrainer.Tag, $"set DatagenCallback: ret({ret})");
        }
~~~
<sampleapp.cs>
~~~
         dataset.AddGenerator(NNTrainerDatasetMode.Train, userData);
         dataset.eventReceived += (s, e) => {       
               //Consider the following array as a two-dimensional like float[numInput, inputLength]            
                e.Input = new float[inputNum * inputLength];
                e.Label = new float[LabelNum * labelLength];
                e.Last = new float[1];
                int idx = 0;
                for (int i = 0; i < inputNum; ++i)
                {
                    for (int j = 0; j < inputLength; j++)
                        e.Input[idx++] = 0.1f; //outVec[0][i] = 0.0f;
                }

                e.Label[0] = 1.0f; //outLabel[0][0] = 1.0f;
                idx = 1;
                for (int i = 0; i < LabelNum; ++i)
                {
                    for (int j = 1; j < labelLength; j++)
                            e.Label[idx++] = 0.0f; //outLabel[i][j] = 0.0f;
                }  
                if (count == 10)
                {
                    e.Last[0] = 1;
                    count = 0;
                }
                else
                {
                    e.Last[0] = 0;
                    count++;
                }
            };
~~~
but this is my mistake.
**input is not a two-dimensional array of contiguous memory space, It is a tree structures.
**input point first *float [input_num] and each *float [input_num] point first address of tensor with c, h and w.

At least int the above structure, it should be operated when input_num is one. but. it is not operated.
The result of debug as below.
nntrainer(c++) does not have value which is set by Marshal.Copy() in callback(c#).

nntrainer call callback and the callback in C# is called, and the data is well put into Args.Input and label in sample app as event invoke. there is no error by Marshal.Copy().
I don't know. Is it because Interop treats **pointer as an IntPtr?
args.Input and args.Input.Length have no issue. input should be pointed first tensor data address of nntrainer.
~~~
Marshal.Copy(args.Input, 0, intput, args.Input.Length);
~~~
Anyway, We need some interface to support multi input_num  and solve double pointer issue.
